tests/recovery/recovery_test/recovery_test.go

```diff
@@ -168,6 +172,172 @@ func TestSnapshotFixInvalidFiles(t *testing.T) {
+	require.NoError(t, err)
+	log.Println("Copy content and modify the files.")
+
+	// kill the kopia command before it exits
+	kopiaExe := os.Getenv("KOPIA_EXE")
+
+	cmd := exec.Command(kopiaExe, "repo", "connect", "filesystem", "--path="+dataRepoPath, "--content-cache-size-mb", "500", "--metadata-cache-size-mb", "500", "--no-check-for-updates")
+	env := []string{"KOPIA_PASSWORD=" + testenv.TestRepoPassword}
+	cmd.Env = append(os.Environ(), env...)
+
+	o, err := cmd.CombinedOutput()
+	require.NoError(t, err)
+	log.Println(string(o))
+
+	cmd = exec.Command(kopiaExe, "snap", "create", copyDir, "--json", "--parallel=1")
+
+	log.Println("Kill the kopia command before it exits:")
+	killOnCondition(t, cmd)
+
+	log.Println("Verify snapshot corruption:")
+	// verify snapshot corruption
+	err = bm.VerifySnapshot()
+	require.NoError(t, err)
+
+	// Create a temporary dir to restore a snapshot
+	restoreDir := t.TempDir()
+	if restoreDir == "" {
+		t.FailNow()
+	}
+
+	// try to restore a snapshot without any error messages.
+	stdout, err := bm.RestoreGivenOrRandomSnapshot(snapID, restoreDir)
+	require.NoError(t, err)
+
+	log.Println(stdout)
+
+	log.Println("Compare restored data and original data:")
+	require.NoError(t, bm.FileHandler.CompareDirs(restoreDir, cmpDir))
+}
+
+func killOnCondition(t *testing.T, cmd *exec.Cmd) {
+	t.Helper()
+
+	stderrPipe, err := cmd.StderrPipe()
+	require.NoError(t, err)
+
+	// execute kill -9 while receive ` | 1 hashing, 0 hashed (65.5 KB), 0 cached (0 B), uploaded 0 B, estimating...` message
+	var wg sync.WaitGroup
+
+	var mu sync.Mutex
+
+	// Add a WaitGroup counter for the first goroutine
+	wg.Add(1)
+
+	errOut := bytes.Buffer{}
+
+	go func() {
+		mu.Lock()
+		defer mu.Unlock()
+		defer wg.Done()
+
+		// Create a scanner to read from stderrPipe
+		scanner := bufio.NewScanner(stderrPipe)
+		for scanner.Scan() {
+			output := scanner.Text()
+			log.Println(output)
+			errOut.Write(scanner.Bytes())
+			errOut.WriteByte('\n')
+
+			log.Println(output)
+			// Check if the output contains the "hashing" etc.
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") || strings.Contains(output, "Snapshotting") {
+				log.Println("Detaching and terminating target process")
+				cmd.Process.Kill()
+
+				break
+			}
+		}
+	}()
+
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		panic(err)
+	}
+
+	wg.Add(1)
+
+	o := bytes.Buffer{}
+
+	go func() {
+		defer wg.Done()
+
+		// Create a scanner to read from stdoutPipe
+		scanner := bufio.NewScanner(stdoutPipe)
+		for scanner.Scan() {
+			output := scanner.Text()
+			log.Println(output)
+			o.Write(scanner.Bytes())
+			o.WriteByte('\n')
+
+			log.Println("snapshot create successfully", output)
+			// Check if the output contains the "copying" text
+			if strings.Contains(output, "hashing") && strings.Contains(output, "hashed") && strings.Contains(output, "uploaded") {
+				cmd.Process.Kill()
+				break
```

Comment:
You should drain the `stdoutPipe` before stopping the goroutine or your subcommand can get deadlocked. If a subprocess is blocked on writing to the std-err before of buffer limitations then you will not be able to kill it and hence your `cmd.Wait()` or `cmd.Run()` will get stuck forever.

Reply:
Since the log messages related to uploading are only printed to the stderrPipe, I believe there is no need to include the stdoutPipe anymore. Therefore, I think there is no requirement to drain the stdoutPipe in this scenario.